clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java

```diff
@@ -347,16 +352,35 @@ private void sendProduceRequests(Map<Integer, List<RecordBatch>> collated, long
      * Create a produce request from the given record batches
      */
     private void sendProduceRequest(long now, int destination, short acks, int timeout, List<RecordBatch> batches) {
+        if (batches.isEmpty())
+            return;
+
         Map<TopicPartition, MemoryRecords> produceRecordsByPartition = new HashMap<>(batches.size());
         final Map<TopicPartition, RecordBatch> recordsByPartition = new HashMap<>(batches.size());
+
+        // find the minimum magic version used when creating the record sets
+        byte minUsedMagic = apiVersions.maxUsableProduceMagic();
+        for (RecordBatch batch : batches) {
+            if (batch.magic() < minUsedMagic)
+                minUsedMagic = batch.magic();
+        }
+
         for (RecordBatch batch : batches) {
             TopicPartition tp = batch.topicPartition;
-            produceRecordsByPartition.put(tp, batch.records());
+            MemoryRecords records = batch.records();
+
+            // down convert if necessary to the minimum magic used. This is intended to handle edge cases around
+            // cluster upgrades where brokers may not all support the same message format version. For example,
+            // if a partition migrates from a broker which is supporting the new magic version to one which doesn't,
+            // then we will need to convert.
```

Comment:
Discussed offline, but for posterity, down-conversion on the broker only helps when the broker supports the new produce request version. However, the client also needs to support older versions of the produce request, which must use older versions of the message format. The difficulty is that there is a delay between the time that the producer starts building the batch and the time that we send the request, and we may have chosen the message format based on out-dated metadata. In the worst case, we optimistically chose to use the new message format, but found that the broker didn't support it, so we need to down-convert on the client before sending.